librz/bin/pdb/pdb.c

```diff
@@ -1,1433 +1,589 @@
-		if (length % page_size) {
-			num_pages++;
-		}
-	}
-	return num_pages;
-}
-
-static int init_pdb7_root_stream(RzPdb *pdb, int *root_page_list, int pages_amount,
-	EStream indx, int root_size, int page_size) {
-	RZ_PDB_STREAM *pdb_stream = 0;
-	int tmp_data_max_size = 0;
-	char *tmp_data = NULL, *data_end;
-	int stream_size = 0;
-	int num_streams = 0;
-	int *sizes = NULL;
-	int num_pages = 0;
-	int data_size = 0;
-	char *data = NULL;
-	int i = 0;
-	int pos = 0;
-
-	RZ_PDB7_ROOT_STREAM *root_stream7;
-
-	pdb->root_stream = RZ_NEW0(RZ_PDB7_ROOT_STREAM);
-	init_r_pdb_stream(&pdb->root_stream->pdb_stream, pdb->buf, root_page_list, pages_amount,
-		indx, root_size, page_size);
-
-	root_stream7 = pdb->root_stream;
-	pdb_stream = &(root_stream7->pdb_stream);
-
-	stream_file_get_size(&pdb_stream->stream_file, &data_size);
-	data = (char *)calloc(1, data_size);
-	if (!data) {
-		return 0;
-	}
-	stream_file_get_data(&pdb_stream->stream_file, data);
-
-	num_streams = *(int *)data;
-	tmp_data = data;
-	tmp_data += 4;
-
-	root_stream7->num_streams = num_streams;
-
-	tmp_data_max_size = (data_size - (num_streams * 4 - 4));
-	data_end = data + tmp_data_max_size;
-	if (tmp_data_max_size > data_size) {
-		RZ_FREE(data);
-		eprintf("Invalid max tmp data size.\n");
-		return 0;
-	}
-	if (num_streams < 0 || tmp_data_max_size <= 0) {
-		RZ_FREE(data);
-		eprintf("Too many streams: current PDB file is incorrect.\n");
-		return 0;
-	}
-
-	sizes = (int *)calloc(num_streams, 4);
-	if (!sizes) {
-		RZ_FREE(data);
-		eprintf("Size too big: current PDB file is incorrect.\n");
-		return 0;
-	}
-
-	for (i = 0; i < num_streams && (tmp_data + 4 < data_end); i++) {
-		stream_size = *(int *)(tmp_data);
-		tmp_data += 4;
-		if (stream_size == UT32_MAX) {
-			stream_size = 0;
-		}
-		memcpy(sizes + i, &stream_size, 4);
-	}
-
-	tmp_data = ((char *)data + num_streams * 4 + 4);
-	root_stream7->streams_list = rz_list_new();
-	RzList *pList = root_stream7->streams_list;
-	SPage *page = 0;
-	for (i = 0; i < num_streams; i++) {
-		num_pages = count_pages(sizes[i], page_size);
-
-		if ((pos + num_pages) > tmp_data_max_size) {
-			RZ_FREE(data);
-			RZ_FREE(sizes);
-			eprintf("Warning: looks like there is no correct values of stream size in PDB file.\n");
-			return 0;
-		}
-
-		ut32 size = num_pages * 4;
-		ut8 *tmp = (ut8 *)calloc(num_pages, 4);
-		page = RZ_NEW0(SPage);
-		if (num_pages != 0) {
-			if ((pos + size) > tmp_data_max_size) {
-				eprintf("Data overrun by num_pages.\n");
-				RZ_FREE(data);
-				RZ_FREE(sizes);
-				RZ_FREE(tmp);
-				RZ_FREE(page);
-				return 0;
-			}
-			memcpy(tmp, tmp_data + pos, num_pages * 4);
-			pos += size;
-			page->stream_size = sizes[i];
-			page->stream_pages = tmp;
-			page->num_pages = num_pages;
-		} else {
-			page->stream_size = 0;
-			page->stream_pages = 0;
-			page->num_pages = 0;
-			free(tmp);
-		}
-
-		rz_list_append(pList, page);
+static char *pdb_type_as_string_regular(const RzTypeDB *db, const RzPdb *pdb, const RzList *types) {
```

Comment:
I feel like this kind of function is dupped already in too many places. Can we use a RzType API instead of this custom one? At the least https://github.com/rizinorg/rizin/pull/1678 should be able to provide the right API, but I believe there should be one API already for this anyway cc @XVilka

Reply:
Yes, just see `rz_type_as_string()` and other related APIs

Reply:
The types we need to print are `RzBaseType`. There is only an IPI about printing `RzBaseType` which belongs to the `tsc` command. So there is dupped indeed but I have no other way.

Reply:
This function does not print the new line so the output is a mess.

Reply:
Can be done once https://github.com/rizinorg/rizin/issues/1775 is solved.